File: ruby-gem/lib/models/physical_product.rb
# frozen_string_literal: true

# Class from DFC Business Ontology: #PhysicalProduct
require_relative '../core/semantic_object'
require_relative 'what_subject'


module DfcLinkmlConnector
  module Models
    class PhysicalProduct < WhatSubject
      SEMANTIC_TYPE = "dfc-b:PhysicalProduct".freeze

      # @return [String]
      attr_accessor :image

      # @return [Float]
      attr_accessor :quantity

      # @return [String]
      attr_accessor :concerned_by

      # @return [String]
      attr_accessor :constitued_by

      # @return [String]
      attr_accessor :consumed_by

      # @return [String]
      attr_accessor :fulfills

      # @return [String]
      attr_accessor :owned_by

      # @return [String]
      attr_accessor :produced_by

      # @return [String]
      attr_accessor :represents

      # @return [String]
      attr_accessor :traced_by

      # @param semanticId [String]
      # @param date: nil, description: nil, name: nil, characteristicOf: nil, dimension: nil, image: nil, quantity: nil, concernedBy: nil, constituedBy: nil, consumedBy: nil, fulfills: nil, ownedBy: nil, producedBy: nil, represents: nil, tracedBy: nil
      def initialize(semanticId, date: nil, description: nil, name: nil, characteristicOf: nil, dimension: nil, image: nil, quantity: nil, concernedBy: nil, constituedBy: nil, consumedBy: nil, fulfills: nil, ownedBy: nil, producedBy: nil, represents: nil, tracedBy: nil)
        super(semanticId, date: date, description: description, name: name, characteristicOf: characteristicOf, dimension: dimension)
        @image = image
        @quantity = quantity
        @concerned_by = concernedBy
        @constitued_by = constituedBy
        @consumed_by = consumedBy
        @fulfills = fulfills
        @owned_by = ownedBy
        @produced_by = producedBy
        @represents = represents
        @traced_by = tracedBy
        self.semanticType = "dfc-b:PhysicalProduct"
        registerSemanticProperty("dfc-b:Image", &method("image")).valueSetter = method("image=")
        registerSemanticProperty("dfc-b:quantity", &method("quantity")).valueSetter = method("quantity=")
        registerSemanticProperty("dfc-b:concernedBy", &method("concerned_by")).valueSetter = method("concerned_by=")
        registerSemanticProperty("dfc-b:constituedBy", &method("constitued_by")).valueSetter = method("constitued_by=")
        registerSemanticProperty("dfc-b:consumedBy", &method("consumed_by")).valueSetter = method("consumed_by=")
        registerSemanticProperty("dfc-b:fulfills", &method("fulfills")).valueSetter = method("fulfills=")
        registerSemanticProperty("dfc-b:ownedBy", &method("owned_by")).valueSetter = method("owned_by=")
        registerSemanticProperty("dfc-b:producedBy", &method("produced_by")).valueSetter = method("produced_by=")
        registerSemanticProperty("dfc-b:represents", &method("represents")).valueSetter = method("represents=")
        registerSemanticProperty("dfc-b:tracedBy", &method("traced_by")).valueSetter = method("traced_by=")
      end

      Core::SemanticObject.type_registry[SEMANTIC_TYPE] = self
    end
  end
end

File: typescript-connector/src/models/PhysicalProduct.ts
// Class from DFC Business Ontology: #PhysicalProduct
import { SemanticObject } from "../core/SemanticObject.js";
import { WhatSubject, type WhatSubjectParams } from "./WhatSubject.js";
import type { Agent } from "./Agent.js";
import type { LocalizedProduct } from "./LocalizedProduct.js";
import type { ProductBatch } from "./ProductBatch.js";
import type { QuantitativeValue } from "./QuantitativeValue.js";

export interface PhysicalProductParams extends WhatSubjectParams {
  image?: string;
  quantity?: number;
  concernedBy?: string;
  constituedBy?: string;
  consumedBy?: string;
  fulfills?: string[];
  producedBy?: string;
  hasQuantity?: QuantitativeValue | string;
  ownedBy?: Agent | string;
  represents?: (LocalizedProduct | string)[];
  tracedBy?: ProductBatch | string;
}
export class PhysicalProduct extends WhatSubject {
  static get SEMANTIC_TYPE(): string {
    return "dfc-b:PhysicalProduct";
  }

  image?: string;
  quantity?: number;
  concernedBy?: string;
  constituedBy?: string;
  consumedBy?: string;
  fulfills?: string[];
  producedBy?: string;
  hasQuantity?: QuantitativeValue | string;
  ownedBy?: Agent | string;
  represents?: (LocalizedProduct | string)[];
  tracedBy?: ProductBatch | string;

  constructor(
    semanticId: string,
    params?: PhysicalProductParams,
  ) {
    super(semanticId, params);
    this.image = params?.image;
    this.quantity = params?.quantity;
    this.concernedBy = params?.concernedBy;
    this.constituedBy = params?.constituedBy;
    this.consumedBy = params?.consumedBy;
    this.fulfills = params?.fulfills;
    this.producedBy = params?.producedBy;
    this.hasQuantity = params?.hasQuantity;
    this.ownedBy = params?.ownedBy;
    this.represents = params?.represents;
    this.tracedBy = params?.tracedBy;
    this.semanticType = PhysicalProduct.SEMANTIC_TYPE;
    this.registerSemanticProperty("dfc-b:Image", () => this.image);
    this.registerSemanticProperty("dfc-b:quantity", () => this.quantity);
    this.registerSemanticProperty("dfc-b:concernedBy", () => this.concernedBy);
    this.registerSemanticProperty("dfc-b:constituedBy", () => this.constituedBy);
    this.registerSemanticProperty("dfc-b:consumedBy", () => this.consumedBy);
    this.registerSemanticProperty("dfc-b:fulfills", () => this.fulfills);
    this.registerSemanticProperty("dfc-b:producedBy", () => this.producedBy);
    this.registerSemanticProperty("dfc-b:hasQuantity", () => this.hasQuantity);
    this.registerSemanticProperty("dfc-b:ownedBy", () => this.ownedBy);
    this.registerSemanticProperty("dfc-b:represents", () => this.represents);
    this.registerSemanticProperty("dfc-b:tracedBy", () => this.tracedBy);
  }
  static {
    SemanticObject.typeRegistry.set(PhysicalProduct.SEMANTIC_TYPE, PhysicalProduct);
  }
}
